fvm/evm/offchain/utils/collection_test.go

```diff
@@ -2,45 +2,252 @@ package utils_test
 
 import (
 	"bufio"
+	"encoding/gob"
 	"encoding/hex"
 	"encoding/json"
+	"fmt"
 	"os"
+	"path/filepath"
 	"strings"
 	"testing"
 
-	"github.com/onflow/cadence"
-	"github.com/onflow/cadence/encoding/ccf"
 	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
 	"github.com/stretchr/testify/require"
 
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/encoding/ccf"
+
 	"github.com/onflow/flow-go/fvm/environment"
 	"github.com/onflow/flow-go/fvm/evm"
 	"github.com/onflow/flow-go/fvm/evm/events"
 	"github.com/onflow/flow-go/fvm/evm/offchain/blocks"
+	"github.com/onflow/flow-go/fvm/evm/offchain/storage"
 	"github.com/onflow/flow-go/fvm/evm/offchain/sync"
 	"github.com/onflow/flow-go/fvm/evm/offchain/utils"
 	. "github.com/onflow/flow-go/fvm/evm/testutils"
-	"github.com/onflow/flow-go/fvm/evm/types"
 	"github.com/onflow/flow-go/model/flow"
 )
 
-func ReplyingCollectionFromScratch(
+func TestTestnetBackwardCompatibility(t *testing.T) {
+	t.Skip("TIME CONSUMING TESTS. Enable the tests with the events files saved in local")
+	// how to run this tests
+	// Note: this is a time consuming tests, so please run it in local
+	//
+	// 1) run the following cli to get the events files across different sporks
+
+	// flow events get A.8c5303eaa26202d6.EVM.TransactionExecuted A.8c5303eaa26202d6.EVM.BlockExecuted
+	//   --start 211176670 --end 211176770 --network testnet --host access-001.devnet51.nodes.onflow.org:9000
+	// > ~/Downloads/events_devnet51_1.jsonl
+	// ...
+	//
+	// 2) comment the above t.Skip, and update the events file paths and checkpoint dir
+	// to run the tests
+	BackwardCompatibleSinceEVMGenesisBlock(
+		t, flow.Testnet, []string{
+			"~/Downloads/events_devnet51_1.jsonl",
+			"~/Downloads/events_devnet51_2.jsonl",
+		},
+		"~/Downloads/",
+		0,
+	)
+}
+
+// BackwardCompatibilityTestSinceEVMGenesisBlock verifies that the offchain package
+// is able to read EVM events from the given file paths and replay blocks since the
+// EVM genesis block and derive a consistant state as the latest onchain EVM state.
+// the eventsFilePaths is a list of file paths that contain ordered EVM events in JSONL format.
+// The EVM events file can be queried by flow cli query, for instance:
+//
+// flow events get A.8c5303eaa26202d6.EVM.TransactionExecuted A.8c5303eaa26202d6.EVM.BlockExecuted
+//   --start 211176670 --end 211176770 --network testnet --host access-001.devnet51.nodes.onflow.org:9000
+//
+// After replaying with each event json file, it will generate a values_<height>.gob and
+// allocators_<height>.gob files as checkpoint, such that when the checkpoint exists, it will loaded
+// and skil replaying the coresponding event json files.
+
```

Comment:
The following part:
```go

// backwardCompatibilityTestSinceEVMGenesisBlock ensures that the offchain package
// can read EVM events from the provided file paths, replay blocks starting from
// the EVM genesis block, and derive a consistent state matching the latest on-chain EVM state.
//
// The parameter `eventsFilePaths` is a list of file paths containing ordered EVM events in JSONL format.
// These EVM event files can be generated using the Flow CLI query command, for example:
//
// flow events get A.8c5303eaa26202d6.EVM.TransactionExecuted A.8c5303eaa26202d6.EVM.BlockExecuted
//
//	--start 211176670 --end 211176770 --network testnet --host access-001.devnet51.nodes.onflow.org:9000
//
// During the replay process, it will generate `values_<height>.gob` and
// `allocators_<height>.gob` checkpoint files for each height. If these checkpoint files exist,
// the corresponding event JSON files will be skipped to optimize replay.
```
Seems to be a duplicated description, as it exists above, with a slightly different wording.